sklearn/decomposition/pca.py

```diff
@@ -367,7 +371,10 @@ def _fit(self, X):
 
         # Handle n_components==None
         if self.n_components is None:
-            n_components = X.shape[1]
+            if self.svd_solver != 'arpack':
+                n_components = min(X.shape)
+            else:
+                n_components = min(X.shape) - 1
```

Comment:
the solver uses [scipy.sparse.linalg.svds](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.svds.html), and according to the docs
the k parameter (n_components) must be such that 1 <= k < min(A.shape).